### PR TITLE
Add CI tests to validate all old provider versions against the current master

### PR DESCRIPTION
**Pull request by paulidale**
- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by slontis** [comment]
I care much less about this than I do about the fips related tests..
If you dont need fips then you should probably just be using the same versions of the providers.

**Comment by paulidale** [comment]
There are additional failures in legacy.

**Comment by slontis** [comment]
There are a few changes to RC4-HMAC 
One of them is this.. https://github.com/openssl/openssl/commit/33219939c782cf363b30e9e899b9997fb1ced440

**Comment by paulidale** [comment]
Most of the issues are fixed.

Refer also: openssl/tools#125 for release procedure update to try to not lose sync.

**Comment by t8m** [comment]
IMO this is too much. I really do not think we want to test all the provider versions ever released. Let's take an example we make some mistake causing regression in a provider in version 3.0.3. We fix it in 3.0.4 adding a test for it. Do we want that test to be special-cased? This in a few years explodes into unmaintainable mess.

I'd just test validated versions of the fips provider plus the latest one from the 3.0 branch and a few selected versions of the legacy provider that are known to not contain any regressions. So again perhaps 3.0.0 plus the latest from the 3.0 branch.

Anyway I would not recommend anyone to use the legacy provider from 3.0.0 especially given the RC4-MD5 security issue fixed in there.

**Comment by mattcaswell** [comment]
There was *extensive* discussion about these very points in the OMC meeting yesterday. The result was this ongoing vote:

https://github.com/openssl/general-policies/issues/27

The argument in favour of doing this is that we have committed to a forward compatible ABI for providers. The only way we know we haven't broken that promise is to test it. Therefore we test it by using our own released providers as a proxy for all 3rd party providers.

> Do we want that test to be special-cased?

Yes. We need to do this special casing anyway even if all you are testing is a few selected prior versions. Adding more versions of providers should not mean adding more special casing, it just means you need to be more accurate about which specific version number of a provider contains the fix for a bug that a test is testing. It should *not* (generally) mean needing to special case more tests that you would otherwise. The only case I can think of where this could occur is where you introduce a new bug and then fix it in between validated versions of the fips provider.

> This in a few years explodes into unmaintainable mess.

We need to do this in any case at least for the validated versions of the FIPS provider. If you are already doing it for the validated versions it really doesn't add to the explosion to test all versions.

> Anyway I would not recommend anyone to use the legacy provider from 3.0.0 especially given the RC4-MD5 security issue fixed in there.

This is misunderstanding the objective. We are not testing the legacy provider from old versions because we necessarily expect or want anyone to use it. We are testing it because it enables us to check that we have not inadvertently broken the provider ABI. The legacy provider is acting as a proxy for other 3rd party providers.

**Comment by slontis** [comment]
This is still kind of reliant on having high test coverage to some extent.

**Comment by t8m** [comment]
I still see no point in running the tests with the intermediate versions of legacy or even of the fips provider. Although you're right that in most cases the version checks in the tests will have to be there anyway, my example with one version regression in a legacy or unvalidated fips provider still holds. And running all these version tests would be just mindless resource consumption. What if there are 20 3.0.x releases and 10 3.1.x releases at some point.

**Comment by mattcaswell** [comment]
> Although you're right that in most cases the version checks in the tests will have to be there anyway, my example with one version regression in a legacy or unvalidated fips provider still holds.

Right, introducing a bug in one release, and fixing it in the next, in between validated versions would introduce a new special case that you wouldn't otherwise have. I don't expect there to be too many of those, so I don't really see this as a problem.

> And running all these version tests would be just mindless resource consumption. What if there are 20 3.0.x releases and 10 3.1.x releases at some point.

I expect this to be refined over time. If we get to the point where we have so many releases that resource consumption is a problem then we will modify the policy to test a representative sample. At the moment we have 5 which is really not that many. We have 1 validated version which is not a sufficiently large sample size, and we're not testing the legacy provider at all - which is a missed opportunity to test provider ABI codepaths that may not be exercised by the fips provider.

**Comment by slontis** [comment]
This also doesnt need to be run very often I would imagine?

**Comment by mattcaswell** [comment]
> This also doesnt need to be run very often I would imagine?

Correct. The in-progress vote only talks about doing it on release. I suspect we may do it a little more frequently than that in practice.

**Comment by t8m** [comment]
There is one more potential issue with testing the legacy provider that might complicate its testing in these scenarios. The legacy provider calls into libcrypto and is not completely isolated from it like the fips provider. This probably is not so big deal if we are testing legacy provider with newer libcrypto (from the same major version). However it won't work universally if the legacy provider is from newer branch than the libcrypto that is being tested. I understand that we currently do not have this problem but eventually we might.

**Comment by t8m** [comment]
> Correct. The in-progress vote only talks about doing it on release. I suspect we may do it a little more frequently than that in practice.

I'd say we might want to run this test on the latest 3.0.x version released on pull requests. This would enable correctly skipping the newly added bug fix testcases if they break. But definitely not the full set.

**Comment by t8m** [comment]
Running the full set daily would be more than enough.

**Comment by paulidale** [comment]
> The argument in favour of doing this is that we have committed to a forward compatible ABI for providers.

We also committed equally to a backwards compatible ABI for (at least FIPS) providers at the same time.  This has not been mentioned anywhere and no actions have resulted.  Personally, I only remember discussing FIPS providers being subject to such forwards and backward compatibility requirements & it makes sense for them given how (hopefully) infrequently validations will occur.

Moreover, **none** of the fixes here are ABI changes.  They are bug fixes, the ABI is unchanged throughout.  Our provider ABI seems to be pretty solid, our implementation still has some bugs.


> If we get to the point where we have so many releases that resource consumption is a problem then we will modify the policy to test a representative sample.

This should be an OTC decision (as has been mentioned elsewhere, the OTC is responsible for ensuring that the tests are run).  However, it's an OMC vote that mandates **all** releases be covered.  Modifying the policy becomes an exercise in bureaucracy.


> I understand that we currently do not have this problem but eventually we might.

I'm expecting a world more hurt from the _test everything always_ approach.


My intention is to schedule the full suite daily or weekly on both 3.0.x and master.  There is no point wasting so many cycles on commits or on PRs.  I will keep the parent's FIPS 3.0.0 provider on PRs -- that makes sense IMO.  We **need** to know if validated FIPS providers aren't usable ASAP.

**Comment by paulidale** [comment]
[New CIs passing](https://github.com/openssl/openssl/actions/runs/3064551765).
Switching to cron based scheduling instead of on PR.

**Comment by levitte** [comment]
> Moreover, **none** of the fixes here are ABI changes. They are bug fixes, the ABI is unchanged throughout. Our provider ABI seems to be pretty solid, our implementation still has some bugs.

Agreed.  But we still have to test this, and I suspect that where there are known bugs in older versions of a provider, we need to insert code to not run the tests that we know trigger that bug, or otherwise be able to handle buggy responses.  That's probably the scarier part, that we might end up with a growing database of such "exceptions".

**Comment by paulidale** [comment]
I still don't think we need to test everything.  My experience here was that **all** of the problems were present in the 3.0.0 providers and that **zero** new problems were introduced later.

Moreover, 3.0.4 was identical to 3.0.5 in terms of problems -- not surprising since 3.0.4 was a broken release and 3.0.5 replaced it very soon after.

This is overkill IMO.  Still, the ongoing OMC vote will decide things once and for all.

**Comment by FdaSilvaYY** [comment]
Typo on a commit message : lergacy provider

**Comment by paulidale** [comment]
Thanks @FdaSilvaYY, fixed.

File: test/recipes/90-test_store.t [hunk 1]
```diff
@@ -32,9 +32,14 @@ my @src_files =
       "apps/server.pem" );
 my @data_files =
     ( "testrsa.msb" );
+
```
Comment (slontis) [inline test/recipes/90-test_store.t:35]: Might be worth tracking down the PR for this one also to be consistent

File: .github/workflows/fips-old.yml [hunk 2]
```diff
@@ -5,7 +5,7 @@
 # in the file LICENSE in the source distribution or at
 # https://www.openssl.org/source/license.html
```
Comment (slontis) [inline .github/workflows/fips-old.yml:7]: Can you put a comment at the top here to describe what this does..

File: .github/workflows/provider-compatibility.yml [hunk 3]
```diff
@@ -0,0 +1,97 @@
+# Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
```
Comment (slontis) [inline .github/workflows/provider-compatibility.yml:7]: A comment here also would be good..

**Comment by t-j-h** [comment]
> > Moreover, **none** of the fixes here are ABI changes. They are bug fixes, the ABI is unchanged throughout. Our provider ABI seems to be pretty solid, our implementation still has some bugs.
> 
> Agreed. But we still have to test this, and I suspect that where there are known bugs in older versions of a provider, we need to insert code to not run the tests that we know trigger that bug, or otherwise be able to handle buggy responses. That's probably the scarier part, that we might end up with a growing database of such "exceptions".

@levitte https://github.com/openssl/general-policies/issues/27 is still waiting on your vote

**Comment by paulidale** [comment]
Vote failed, no point leaving this hanging around.

**Pull request closed without merging**